### PR TITLE
[Foreach Reduction] Use `OpMathType` tensor for intermediate results

### PR DESCRIPTION
**Pull request by crcrpar**
Follow-up of https://github.com/pytorch/pytorch/pull/62646

In APEX, multi_tensor_norm only supports float and half and the dtype of `output` and `output_per_tensor` is hardcoded as single-precision (see https://github.com/NVIDIA/apex/blob/ae757634efa26a4ed852324f1d32f2159774997b/csrc/multi_tensor_l2norm_kernel.cu#L318).

But in my previous PR, any tensor created in the kernel has the same dtype as the input tensors.
I'm not quite sure why I didn't see any failures in the previous PR but internal math should be performed in 32 bits for 16-bit tensors, in my opinion.

rel: https://github.com/pytorch/pytorch/issues/58833

cc @ptrblck @mcarilli @ngimel

### CODE REVIEW

**Comment by ngimel** [comment]
Does this result in a different behavior between `slow` and mta implementation? By default `norm` is producing same dtype output as input.

**Comment by crcrpar** [comment]
Sorry naming is bad and confusing but output tensors of `torch._foreach_norm` still have the same dtype as input tensors.
The `output_per_tensor` (and `output`) in the description refer to https://github.com/pytorch/pytorch/blob/f346feb06532eb8c2124e8819730fb07eb5bc5b4/aten/src/ATen/native/cuda/ForeachReduceOp.cu#L125, and is used to store the intermediate results like sum of squared in https://github.com/pytorch/pytorch/blob/f346feb06532eb8c2124e8819730fb07eb5bc5b4/aten/src/ATen/native/cuda/ForeachReduceOp.cu#L69.
Currently the output of `torch._foreach_norm` is `ret_per_tensor` in https://github.com/pytorch/pytorch/blob/f346feb06532eb8c2124e8819730fb07eb5bc5b4/aten/src/ATen/native/cuda/ForeachReduceOp.cu#L126.

**Comment by ngimel** [comment]
Can you please come up with a test case that would be triggered by not using opmath_t (e.g. overflow before taking sqrt) and add it here?

File: aten/src/ATen/OpMathType.h [hunk 1]
```diff
@@ -13,4 +14,16 @@ template<> struct OpMathType<at::BFloat16> { using type = float; };
 template<typename T>
 using opmath_type = typename OpMathType<T>::type;
 
+namespace {
+
+c10::ScalarType toOpMathType(const c10::ScalarType scalar_type) {
+  if (scalar_type == c10::ScalarType::Half || scalar_type == c10::ScalarType::BFloat16) {
```
[the remark this answers is not in the document]
Comment (crcrpar) [reply aten/src/ATen/OpMathType.h]: aa6aff8 accesses this :)

**Comment by crcrpar** [comment]
I updated this PR (sorry I've forgotten this for a while)

**Comment by ngimel** [comment]
Is this ready for land?

**Review by ngimel — APPROVED** [review]

**Pull request closed without merging**